flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/StopWithSavepointContextTest.java

```diff
@@ -0,0 +1,249 @@
+
+        testInstance = new StopWithSavepointContext(jobGraph.getJobID(), scheduler, this.log);
+    }
+
+    @Test
+    public void testHappyPathWithSavepointCreationBeforeTermination() throws Exception {
+        assertHappyPath(
+                (savepointPath) -> {
+                    testInstance.handleSavepointCreation(savepointPath, null);
+                    testInstance.handleExecutionTermination(
+                            Collections.singletonList(ExecutionState.FINISHED));
+                });
+    }
+
+    @Test
+    public void testHappyPathWithSavepointCreationAfterTermination() throws Exception {
+        assertHappyPath(
+                (savepointPath) -> {
+                    testInstance.handleExecutionTermination(
+                            Collections.singletonList(ExecutionState.FINISHED));
+                    testInstance.handleSavepointCreation(savepointPath, null);
+                });
+    }
+
+    private void assertHappyPath(Consumer<String> stopWithSavepointCompletion) throws Exception {
+        final String savepointPath = "savepoint-path";
+
+        stopWithSavepointCompletion.accept(savepointPath);
+
+        assertThat(testInstance.getResult().get(), is(savepointPath));
+
+        // the happy path won't restart the CheckpointCoordinator
+        assertCheckpointSchedulingDisabled();
+    }
+
+    @Test
+    public void testSavepointCreationFailure() {
+        final Exception exception = new Exception("Expected exception during savepoint creation.");
+        testInstance.handleSavepointCreation(null, exception);
+
+        try {
+            testInstance.getResult().get();
+            fail("An ExecutionException is expected.");
+        } catch (Throwable e) {
+            final Optional<Throwable> actualException =
+                    ExceptionUtils.findThrowableWithMessage(e, exception.getMessage());
+            assertTrue(actualException.isPresent());
+        }
+
+        assertCheckpointSchedulingEnabled();
+    }
+
+    @Test
+    public void testNoTerminationHandlingAfterSavepointCompletion() throws Exception {
+        assertNoTerminationHandling(
+                () -> {
+                    testInstance.handleSavepointCreation("savepoint-path", null);
+                    testInstance.handleExecutionTermination(
+                            // the task failed and was restarted
+                            Collections.singletonList(ExecutionState.RUNNING));
+                });
+    }
+
+    @Test
+    public void testNoTerminationHandlingBeforeSavepointCompletion() throws Exception {
+        assertNoTerminationHandling(
+                () -> {
+                    testInstance.handleExecutionTermination(
+                            // the task failed and was restarted
+                            Collections.singletonList(ExecutionState.RUNNING));
+                    testInstance.handleSavepointCreation("savepoint-path", null);
+                });
+    }
```

Comment:
What happens if `handleSavepointCreation(null, throwable)`?

Reply:
The contract was described properly in the new implementation. Corresponding tests were added to `StopWithSavepointTerminationHandlerImplTest`.